src/main/java/frc/robot/subsystems/drivetrain/SwerveDrive.java

```diff
@@ -0,0 +1,264 @@
+        rotation *= Constants.SwerveDrive.ROTATION_MULTIPLIER;
+        // multiplies the 2D rotation matrix by the robot heading, there by rotating the coordinate system
+        // see https://en.wikipedia.org/wiki/Rotation_matrix
+        double[][] rotationMat = {{Math.cos(robotAngle), -Math.sin(robotAngle)},
+                {Math.sin(robotAngle), Math.cos(robotAngle)}};
+        double[] speeds = Utils.matrixVectorMult(rotationMat, new double[]{forward, strafe});
+        System.out.println("forward" + forward);
+        System.out.println("strafe" + strafe);
+
+        // if the drive style is field oriented, changes the forward and strafe to be according to the field coordinate system
+        if (isFieldOriented) {
+            forward = speeds[0];
+            strafe = speeds[1];
+        }
+        System.out.println("relative forward" + speeds[0]);
+        System.out.println("relative strafe" + speeds[1]);
+
+        double[] robotHeading = {forward, strafe, rotation};
+
+        return robotHeading;
+    }
+
+    /**
+     * calculates the velocity vector of each wheel
+     *
+     * @param robotHeading the three joystick outputs:
+     *                     forward the heading of the robot in the Y direction
+     *                     strafe the heading of the robot in the X direction
+     *                     rotation the rotation of the robot
+     * @return an array of length 8 in which each pair is the X and Y velocities of each wheel
+     */
+    public static double[] calculateWheelVelocities(double[] robotHeading) {
+
+
+        // multiplies M by the robotHeading to obtain the wheel velocities
+        double[] wheelVelocities = Utils.matrixVectorMult(M, robotHeading);
+
+        return wheelVelocities;
+
+    }
+
+    /**
+     * set the angle of the wheels on the robot to lock the robot in place
+     */
+    public void lock() {
+        // calculates the lock angles of the wheels
+        double[] lockAngles = calculateLockAngles();
+
+        for (int i = 0; i < 4; i++) {
+            swerveModules[i].setSpeed(0);
+            if (i == 2) {
+                swerveModules[i].setAngle(lockAngles[i + 1]);
+            }
+            if (i == 3) {
+                swerveModules[i].setAngle(lockAngles[i - 1]);
+            }
+        }
+    }
+
+    /**
+     * calculates the angles for which the wheels will lock in place
+     */
+    public double[] calculateLockAngles() {
+        double[] lockAngles = new double[4];
+
+        for (int i = 0; i < 4; i++) {
+            lockAngles[i] = Math.PI / 2 - Math.atan(Constants.SwerveDrive.ROBOT_WIDTH / Constants.SwerveDrive.ROBOT_LENGTH) + i * Math.PI / 2;
+        }
+
+        return lockAngles;
+    }
+
+
+    /**
+     * stops all the wheels
+     */
+    public void stop() {
+        for (int i = 0; i < swerveModules.length; i++) {
+            swerveModules[i].setSpeed(0);
+            swerveModules[i].stopAngleMotor();
```

Comment:
Use a "for-each" loop:
```suggestion
        for (SwerveModule swerveModule : swerveModules) {
            swerveModule.setSpeed(0);
            swerveModule.stopAngleMotor();
```